src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -662,11 +815,49 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
             pat_constructors(cx, row[0], pcx).unwrap_or(vec![])
         }).collect();
         debug!("used_ctors = {:#?}", used_ctors);
-        let all_ctors = all_constructors(cx, pcx);
+        // `all_ctors` are all the constructors for the given type, which
+        // should all be represented (or caught with the wild pattern `_`).
+        // `value_constructors` is true if we may exhaustively consider all
+        // the possible values (e.g. integers) of a type as its constructors.
+        let (all_ctors, value_constructors) = all_constructors(cx, pcx);
```

Comment:
cc @nikomatsakis on the usage of `value_constructors`, I'm not sure if it's necessary or desireable.

Reply:
The reason I've done it this way is because the other methods for matching only permit exhaustive matching on enums, and so they suggest using `_` for integers. Although it might be possibly to modify the other methods (though I couldn't spot how to do it simply), I think it's far more straightforward to have a separate case for types that can be constructed this way using `value_constructors`.

Reply:
Why not check the type and feature-gate again?

Reply:
I could do that instead.

Reply:
I think it would allow more local reasoning (we want diagnostic logic to be separated as much as possible from the checking logic).

Reply:
Maybe it'd be reasonable to simply feature-gate and treat any constructor corresponding to a set of ranges as being amenable to interval exhaustiveness checking. There's no reason this approach shouldn't work in general for values with range-based constructors, so we can eliminate this explicit flagging while still preserving the more precise semantics for ranges.